File: Xahla-API-Core/src/xahla/context/objects/OrthographicCamera.java
package xahla.context.objects;

import org.joml.Matrix4f;
import org.joml.Vector2f;
import org.joml.Vector2i;
import org.joml.Vector3f;

import xahla.client.graphics.Graphics;
import xahla.context.ClientContext;
import xahla.context.components.OrthographicProjection;
import xahla.core.ObjectPriority;
import xahla.core.Priority;

/**
 * The Orthographic Camera contains an orthographic projection for 2D rendering.
 * 
 * @author Cochetooo
 * @version 1.1
 */
@ObjectPriority(priority=Priority.HIGH)
public class OrthographicCamera extends Camera {

	/**
	 * Add an Orthographic Camera to the Context and initialize it with the Orthographic configuration (orthographic.json).
	 * @param name		The name of the Camera. (the one by default is "MainCamera").
	 * @param Context	The context of the program as a Client Context.
	 */
	public OrthographicCamera(String name, ClientContext Context) {
		super(name, Context);
		
		String[] posData = Context.getConfigString("Orthographic", "position").split("-");
		String[] orthoData = Context.getConfigString("Orthographic", "ortho").split("-");
		
		float[] ortho = new float[4];
		Vector2i dim = this.getContext().getWindow().getWindowDimension();
		
		for (int i = 0; i < 3; i++) {
			if (orthoData[i].equals("width"))
				ortho[i] = dim.x;
			else if (orthoData[i].equals("height"))
				ortho[i] = dim.y;
			else if (orthoData[i].equals("auto")) {
				switch (i) {
				case 0, 3 -> ortho[i] = 0;
				case 1 -> ortho[i] = Graphics.screenDimension().x;
				case 2 -> ortho[i] = Graphics.screenDimension().y;
				}
			} else {
				ortho[i] = Float.parseFloat(orthoData[i]);
			}
		}
		
		this.transform().setLocalPosition(new Vector3f(Float.parseFloat(posData[0]), Float.parseFloat(posData[1]), 0));
		this.add(new OrthographicProjection(this, ortho));
	}
	
	/**
	 * @param name		The name of the Camera. (the one by default is "MainCamera").
	 * @param Context	The context of the program as a Client Context.
	 * @param pos		The position of the Camera in the world.
	 * @param size		The size of the projection (top, left, right, bottom).
	 */
	public OrthographicCamera(String name, ClientContext Context, Vector2f pos, float[] size) {
		super(name, new Vector3f(pos.x, pos.y, 0), Context);
		this.add(new OrthographicProjection(this, size));
	}
	
	/** @return The Projection Matrix calculated in the Orthographic Projection component. */
	public Matrix4f projectionMatrix() { return ((OrthographicProjection) this.getComponent("OrthographicProjection")).getProjection(); }

}

File: Render/src/main/kotlin/objects/OrthographicCamera.kt
package objects

import UseComponent
import XHR_SCREEN_DIMENSION
import clientContext
import components.COrthographicProjection
import config
import org.joml.Vector3f

class OrthographicCamera(name: String) : Entity(name) {

    @UseComponent val projection: COrthographicProjection

    init {
        val posData = (config("gl.position") as String).split("-")
        val orthoData = (config("gl.ortho") as String).split("-")

        val ortho = FloatArray(4)
        val dim = clientContext().window.dimension

        for (i in 0 until 4) {
            if (orthoData[i] == "width")
                ortho[i] = dim.x.toFloat()
            else if (orthoData[i] == "height")
                ortho[i] = dim.y.toFloat()
            else if (orthoData[i] == "auto") {
                when (i) {
                    0, 3    -> ortho[i] = 0.0f
                    1       -> ortho[i] = XHR_SCREEN_DIMENSION.x
                    2       -> ortho[i] = XHR_SCREEN_DIMENSION.y
                }
            } else {
                ortho[i] = orthoData[i].toFloat()
            }
        }

        this.detached = true
        this.projection = COrthographicProjection(this, ortho)
        this.transform.position.set(Vector3f(posData[0].toFloat(), posData[1].toFloat(), 0.0f))
    }

}